magento/magento2
# Cli info di

Added cli command to retrieve DI configured for the certain class or interface: which preference is defined, which virtual types exist, how parameters are configured, what plugins exist. 
### Description


### Fixed Issues (if relevant)

1.magento-hackathon/Imagine2017#11: Create new CLI command: Collect Dependency Injection Info for the Interfaces and Classes

### Manual testing scenarios

1. Run ``bin/magento dev:di:info "Magento\Catalog\Api\Data\ProductInterface"``
2. Expected Result:
```
DI configuration for the class Magento\Catalog\Api\Data\ProductInterface in the GLOBAL area

Preference: Magento\Catalog\Model\Product

Constructor Parameters:
+--------------------------------+--------------------------------------------------------------------------+------------------+
| Name                           | Requested Type                                                           | Configured Value |
+--------------------------------+--------------------------------------------------------------------------+------------------+
| context                        | Magento\Framework\Model\Context                                          |                  |
| registry                       | Magento\Framework\Registry                                               |                  |
| extensionFactory               | Magento\Framework\Api\ExtensionAttributesFactory                         |                  |
| customAttributeFactory         | Magento\Framework\Api\AttributeValueFactory                              |                  |
| storeManager                   | Magento\Store\Model\StoreManagerInterface                                |                  |
| metadataService                | Magento\Catalog\Api\ProductAttributeRepositoryInterface                  |                  |
| url                            | Magento\Catalog\Model\Product\Url                                        |                  |
| productLink                    | Magento\Catalog\Model\Product\Link                                       |                  |
| itemOptionFactory              | Magento\Catalog\Model\Product\Configuration\Item\OptionFactory           |                  |
| stockItemFactory               | Magento\CatalogInventory\Api\Data\StockItemInterfaceFactory              |                  |
| catalogProductOptionFactory    | Magento\Catalog\Model\Product\OptionFactory                              |                  |
| catalogProductVisibility       | Magento\Catalog\Model\Product\Visibility                                 |                  |
| catalogProductStatus           | Magento\Catalog\Model\Product\Attribute\Source\Status                    |                  |
| catalogProductMediaConfig      | Magento\Catalog\Model\Product\Media\Config                               |                  |
| catalogProductType             | Magento\Catalog\Model\Product\Type                                       |                  |
| moduleManager                  | Magento\Framework\Module\Manager                                         |                  |
| catalogProduct                 | Magento\Catalog\Helper\Product                                           |                  |
| resource                       | Magento\Catalog\Model\ResourceModel\Product                              |                  |
| resourceCollection             | Magento\Catalog\Model\ResourceModel\Product\Collection                   |                  |
| collectionFactory              | Magento\Framework\Data\CollectionFactory                                 |                  |
| filesystem                     | Magento\Framework\Filesystem                                             |                  |
| indexerRegistry                | Magento\Framework\Indexer\IndexerRegistry                                |                  |
| productFlatIndexerProcessor    | Magento\Catalog\Model\Indexer\Product\Flat\Processor                     |                  |
| productPriceIndexerProcessor   | Magento\Catalog\Model\Indexer\Product\Price\Processor                    |                  |
| productEavIndexerProcessor     | Magento\Catalog\Model\Indexer\Product\Eav\Processor                      |                  |
| categoryRepository             | Magento\Catalog\Api\CategoryRepositoryInterface                          |                  |
| imageCacheFactory              | Magento\Catalog\Model\Product\Image\CacheFactory                         |                  |
| entityCollectionProvider       | Magento\Catalog\Model\ProductLink\CollectionProvider                     |                  |
| linkTypeProvider               | Magento\Catalog\Model\Product\LinkTypeProvider                           |                  |
| productLinkFactory             | Magento\Catalog\Api\Data\ProductLinkInterfaceFactory                     |                  |
| productLinkExtensionFactory    | Magento\Catalog\Api\Data\ProductLinkExtensionFactory                     |                  |
| mediaGalleryEntryConverterPool | Magento\Catalog\Model\Product\Attribute\Backend\Media\EntryConverterPool |                  |
| dataObjectHelper               | Magento\Framework\Api\DataObjectHelper                                   |                  |
| joinProcessor                  | Magento\Framework\Api\ExtensionAttribute\JoinProcessorInterface          |                  |
| data                           |                                                                          | <empty array>    |
+--------------------------------+--------------------------------------------------------------------------+------------------+

Plugins:
+--------+--------+------+
| Plugin | Method | Type |
+--------+--------+------+

Plugins for the Preference:
+--------------------------------------------------------+--------------------+-------+
| Plugin                                                 | Method             | Type  |
+--------------------------------------------------------+--------------------+-------+
| Magento\Swatches\Model\Plugin\Product                  | getMediaAttributes | after |
| Magento\CatalogInventory\Model\Plugin\AfterProductLoad | load               | after |
| Magento\Bundle\Model\Plugin\Product                    | getIdentities      | after |
+--------------------------------------------------------+--------------------+-------+
```


### Contribution checklist
